core/src/main/java/org/infinispan/protostream/impl/TagWriterImpl.java

```diff
@@ -915,4 +931,276 @@ void flush() throws IOException {
+         // Note this will be overwritten if not all 1 - 127 characters below
+         writeVarInt32Direct(startPos + varIntTagLen, strlen);
+
+         int localPos = startPos + varIntTagLen + varIntBytesLen;
+
+         int c;
+         int i;
+         for (i = 0; i < strlen; i++) {
+            c = s.charAt(i);
+            if (c > 127) break;
+
+            out.write(localPos++, (byte) c);
+         }
+
+         out.setPosition(localPos);
+         // All single byte characters
+         if (i == strlen)
+            return;
+
+         byte[] utf8buffer = s.getBytes(StandardCharsets.UTF_8);
+         varIntBytesLen = varIntBytes(utf8buffer.length);
+         out.ensureCapacity(startPos + varIntTagLen + varIntBytesLen + utf8buffer.length);
+
+         writeVarInt32Direct(startPos + varIntTagLen, utf8buffer.length);
+         out.write(startPos + varIntTagLen + varIntBytesLen, utf8buffer);
+         out.setPosition(startPos + varIntTagLen + varIntBytesLen + utf8buffer.length);
+      }
+
+      private int writeFixed32Direct(int i, int value) throws IOException {
+         out.write(i++, (byte) (value & 0xFF));
+         out.write(i++, (byte) ((value >> 8) & 0xFF));
+         out.write(i++, (byte) ((value >> 16) & 0xFF));
+         out.write(i++, (byte) ((value >> 24) & 0xFF));
+         return i;
+      }
+
+      private int writeFixed64Direct(int i, long value) throws IOException {
+         out.write(i++, (byte) (value & 0xFF));
+         out.write(i++, (byte) ((value >> 8) & 0xFF));
+         out.write(i++, (byte) ((value >> 16) & 0xFF));
+         out.write(i++, (byte) ((value >> 24) & 0xFF));
+         out.write(i++, (byte) ((int) (value >> 32) & 0xFF));
+         out.write(i++, (byte) ((int) (value >> 40) & 0xFF));
+         out.write(i++, (byte) ((int) (value >> 48) & 0xFF));
+         out.write(i++, (byte) ((int) (value >> 56) & 0xFF));
+         return i;
+      }
+
+      private int writeVarInt32Direct(int i, int value) throws IOException {
+         while (true) {
+            if ((value & 0xFFFFFF80) == 0) {
+               out.write(i++, (byte) value);
+               break;
+            } else {
+               out.write(i++, (byte) (value & 0x7F | 0x80));
+               value >>>= 7;
+            }
+         }
+         return i;
+      }
+
+      private int writeVarInt64Direct(int i, long value) throws IOException {
+         while (true) {
+            if ((value & 0xFFFFFFFFFFFFFF80L) == 0) {
+               out.write(i++, (byte) value);
+               break;
+            } else {
+               out.write(i++, (byte) ((int) value & 0x7F | 0x80));
+               value >>>= 7;
+            }
+         }
+         return i;
+      }
+
+      private int varIntBytes(long value) {
+         int i = 1;
+         while (true) {
+            if ((value & 0xFFFFFF80L) == 0) {
+               return i;
+            } else {
+               i++;
+               value >>>= 7;
+            }
+         }
+      }
+   }
```

Comment:
nitpicking, can be static and without the if branch

```suggestion
      private static int varIntBytes(long value) {
         int i = 1;
         while ((value & 0xFFFFFFFFFFFFFF80L) != 0) {
            ++i;
            value >>>= 7;
         }
         return i;
      }
```